src/frontend/js/components/SaleTunnel/GenericPaymentButton/index.tsx

```diff
@@ -0,0 +1,303 @@
+  },
+  errorFullProduct: {
+    defaultMessage: 'There are no more places available for this product.',
+    description:
+      'Error message shown when payment creation request failed because there is no remaining available seat for the product.',
+    id: 'components.PaymentButton.errorFullProduct',
+  },
+  errorAddress: {
+    defaultMessage: 'You must have a billing address.',
+    description: "Error message shown when the user didn't select a billing address.",
+    id: 'components.PaymentButton.errorAddress',
+  },
+  errorTerms: {
+    defaultMessage: 'You must accept the terms.',
+    description: "Error message shown when the user didn't check the terms checkbox.",
+    id: 'components.PaymentButton.errorTerms',
+  },
+  pay: {
+    defaultMessage: 'Pay {price}',
+    description: 'CTA label to proceed to the payment of the product',
+    id: 'components.PaymentButton.pay',
+  },
+  payInOneClick: {
+    defaultMessage: 'Pay in one click {price}',
+    description: 'CTA label to proceed to the one click payment of the product',
+    id: 'components.PaymentButton.payInOneClick',
+  },
+  paymentInProgress: {
+    defaultMessage: 'Payment in progress',
+    description: 'Label for screen reader when a payment is in progress.',
+    id: 'components.PaymentButton.paymentInProgress',
+  },
+});
+
+type PaymentInfo = Joanie.Payment & { order_id: string };
+export type OneClickPaymentInfo = Joanie.PaymentOneClick & { order_id: string };
+
+enum ComponentStates {
+  IDLE = 'idle',
+  LOADING = 'loading',
+  ERROR = 'error',
+}
+
+interface Props {
+  buildOrderPayload: (
+    payload: Pick<OrderCreationPayload, 'product_id' | 'has_consent_to_terms'>,
+  ) => OrderCreationPayload;
+}
+
+export const GenericPaymentButton = ({ buildOrderPayload }: Props) => {
+  const intl = useIntl();
+  const API = useJoanieApi();
+  const timeoutRef = useRef<NodeJS.Timeout>();
+  const {
+    eventKey,
+    order,
+    billingAddress,
+    creditCard,
+    product,
+    onPaymentSuccess,
+    props: saleTunnelProps,
+  } = useSaleTunnelContext();
+  const { methods: orderMethods } = useOrders(undefined, { enabled: false });
+  const [payment, setPayment] = useState<PaymentInfo | OneClickPaymentInfo>();
+  const [state, setState] = useState<ComponentStates>(ComponentStates.IDLE);
+  const [error, setError] = useState<PaymentErrorMessageId>(PaymentErrorMessageId.ERROR_DEFAULT);
+  const isMobile = useMatchMediaLg();
+
+  // This pattern is ugly but I couldn't find a better way to achieve it in a nicer way.
+  // Without this, when we call onPaymentSuccess() directly from the after polling function,
+  // it is not the latest version of the function, so the value of `order` inside the context function (defined in GenericSaleTunnel.tsx)
+  // will be undefined, then calling onFinish(order) with an undefined order.
+  const onPaymentSuccessRef = useRef(onPaymentSuccess);
+  onPaymentSuccessRef.current = onPaymentSuccess;
```

Comment:
What about moving most part of `onPaymentSuccess` here ?
We could have this function define here with : 
```ts
WebAnalyticsAPIHandler()?.sendCourseProductEvent(
          CourseProductEvent.PAYMENT_SUCCEED,
          props.eventKey,
        );
        // Once the user has completed the purchase, we need to refetch the orders
        // to update the ordersQuery cache
        invalidateOrders();
        refetchOmniscientOrders();
        queryClient.invalidateQueries({ queryKey: ['user', 'enrollments'] });
        props.onFinish?.(order!);
```
GenericSaleTunnel would just keep the step change.

Then you already retrieve the new order in this component(GenericPaymentButton), it could be used for `onFinish()`